google-cloud-firestore/lib/google/cloud/firestore/query.rb

```diff
@@ -636,13 +890,123 @@ def default_filter
         end
 
         def order_direction direction
-          if direction.to_s.downcase.start_with? "a"
-            :ASCENDING
-          elsif direction.to_s.downcase.start_with? "d"
-            :DESCENDING
-          else
-            :DIRECTION_UNSPECIFIED
+          return :DESCENDING if direction.to_s.downcase.start_with? "d".freeze
+          :ASCENDING
+        end
+
+        def query_has_cursors?
+          query.start_at || query.end_at
+        end
+
+        def values_to_cursor values, query
+          if values.count == 1 && values.first.is_a?(DocumentSnapshot)
+            return snapshot_to_cursor(values.first, query)
+          end
+
+          # pair values with their field_paths to ensure correct formatting
+          order_field_paths = order_by_field_paths query
+          if values.count > order_field_paths.count
+            # raise if too many values provided for the cursor
+            raise ArgumentError, "too many values"
+          end
+
+          values = values.zip(order_field_paths).map do |value, field_path|
+            if field_path == doc_id_path && !value.is_a?(DocumentReference)
+              value = document_reference value
+            end
+            Convert.raw_to_value value
+          end
+
+          Google::Firestore::V1beta1::Cursor.new values: values
+        end
+
+        def snapshot_to_cursor snapshot, query
+          # first, add any inequality filters missing from existing order_by
+          ensure_inequality_field_paths_in_order_by! query
+
+          # second, make sure __name__ is present in order_by
+          ensure_document_id_in_order_by! query
+
+          # lastly, create cursor for all field_paths in order_by
+          values = order_by_field_paths(query).map do |field_path|
+            if field_path == doc_id_path
+              snapshot.ref
+            else
+              snapshot[field_path]
+            end
           end
+
+          values_to_cursor values, query
+        end
+
+        def ensure_inequality_field_paths_in_order_by! query
+          inequality_paths = inequality_filter_field_paths query
+          orig_order = order_by_field_paths query
+
+          inequality_paths.reverse.each do |field_path|
+            next if orig_order.include? field_path
```

Comment:
Note that there should only ever be one `inequality_paths` and hence the direction doesn't matter here. You could even error if there is more than one.

Reply:
Agreed that it could raise. I don't see this being raised in some of the other client implementations though, so we would rely on the API to give an error for multiple inequalities.

I implemented it this way so it could be more declarative and less imperative.